craft_parts/state_manager/step_state.py

```diff
@@ -0,0 +1,124 @@
+    data collected at step run time. Those properties are used to decide whether
+    the step should run again on a new lifecycle execution.
+    """
+
+    part_properties: Dict[str, Any] = {}
+    project_options: Dict[str, Any] = {}
+    files: Set[str] = set()
+    directories: Set[str] = set()
+
+    class Config:
+        """Pydantic model configuration."""
+
+        validate_assignment = True
+        extra = "ignore"
+        allow_mutation = False
+        alias_generator = lambda s: s.replace("_", "-")  # noqa: E731
+        allow_population_by_field_name = True
+
+    @abstractmethod
+    def properties_of_interest(self, part_properties: Dict[str, Any]) -> Dict[str, Any]:
+        """Return relevant properties concerning this step."""
+
+    @abstractmethod
+    def project_options_of_interest(
+        self, project_options: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        """Return relevant project options concerning this step."""
+
+    def diff_properties_of_interest(self, other_properties: Dict[str, Any]) -> Set[str]:
+        """Return properties of interest that differ.
+
+        Take a dictionary of properties and compare to our own, returning
+        the set of property names that are different. Both dictionaries
+        are filtered prior to comparison, only relevant properties are
+        compared.
+
+        :param other_properties: The properties to compare to the
+            project options stored in this state.
+        """
+        return _get_differing_keys(
+            self.properties_of_interest(self.part_properties),
+            self.properties_of_interest(other_properties),
+        )
+
+    def diff_project_options_of_interest(
+        self, other_project_options: Dict[str, Any]
+    ) -> Set[str]:
+        """Return project options that differ.
+
+        Take a dictionary of project_options and compare to our own,
+        returning the set of project option names that are different. Both
+        dictionaries are filtered prior to comparison, only relevant
+        options are compared.
+
+        :param other_project_options: The project options to compare to
+           the project options stored in this state.
+        """
+        return _get_differing_keys(
+            self.project_options_of_interest(self.project_options),
+            self.project_options_of_interest(other_project_options),
+        )
+
+    def marshal(self) -> Dict[str, Any]:
+        """Create a dictionary containing the part state data.
+
+        :return: The newly created dictionary.
+        """
+        return self.dict(by_alias=True)
+
+    def write(self, filepath: Path) -> None:
+        """Write this state to disk."""
+        filepath.parent.mkdir(parents=True, exist_ok=True)
+        yaml_data = self.yaml(by_alias=True)
+        filepath.write_text(yaml_data)
+
+
+def _get_differing_keys(dict1: Dict[str, Any], dict2: Dict[str, Any]) -> Set[str]:
```

Comment:
_get_differing_keys can be simplified to:

```
diff_items = (dict1.items() - dict2.items()) | (dict1.items() - dict2.items())`
return [i[o] for i in diff_items]
```

Reply:
The current implementation (that comes from snapcraft) handles items with value `None` as the equivalent of not having the key declared. I'll check the consequences of comparing without this rule and update if it works in our scenarios.

Reply:
Since this call is always used after it passes the properties/options of interest filter, the `None` special case doesn't really matter and the implementation can be simplified as suggested.